designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,299 @@
+It's used in ESLint core already, in [`lib/shared/ajv.js`](https://github.com/eslint/eslint/blob/22a558228ff98f478fa308c9ecde361acc4caf20/lib/shared/ajv.js#L21).
+Some core ESLint rules describe `default:` values in that schema.
+
+Having two places to describe `default` values can be confusing for developers.
+It can be unclear where a value comes from when its default may be specified in two separate locations.
+This RFC proposes removing the `useDefaults: true` setting in the next major version of ESLint.
+
+### Out of Scope
+
+#### TypeScript Types
+
+Right now, there are two community strategies for declaring rule options types, both with drawbacks:
+
+- [`@types/eslint`](https://npmjs.com/package/@types/eslint) (the community-authored package describing ESLint's types for TypeScript users) [declares `RuleContext`'s `options` as `any[]`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ca231578e79ab1c2f01a308615e60a432061c97a/types/eslint/index.d.ts#L703), indicating it can be an array of any value types.
+  - Drawback: `any` is generally considered a bad practice for TypeScript users as it provides no type descriptions or safety checks.
+- [`@typescript-eslint/utils` > `RuleCreator`](https://typescript-eslint.io/developers/custom-rules#rulecreator) (the recommended TypeScript-enabled rule development utility) declares a [generic `RuleContext<TMessageIds, TOptions>` type with an `options: TOptions` property](https://github.com/typescript-eslint/typescript-eslint/blob/72149e1fe8f57fd217dfc59295a51df68187aad4/packages/utils/src/ts-eslint/Rule.ts#L175).
+  - Drawback: `RuleContext` doesn't factor in root-level `defaultOptions` to the type system (https://github.com/typescript-eslint/typescript-eslint/issues/5439)
+
+This change is limited to the ESLint core repository, which doesn't provide TypeScript types.
+It is out of scope for this RFC to apply `RuleCreator`-style types to `@types/node`, remove `default` from typescript-eslint's schema types, and/or resolve typescript-eslint's factoring default options issue.
+
+## Documentation
+
+We'll want to augment the following docs generators in:
+
+- The core ESLint docs: automating and therefore standardizing how rule docs pages phrase default values
+- typescript-eslint's docs: the same as the core ESLint docs
+- [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator)
+
+We'll also want to mention this in the ESLint core custom rule documentation.
+
+## Drawbacks
+
+- This increases both conceptual and implementation complexity around built-in rule options logic
+  - This RFC believes that enough rules were already implemented with ad-hoc logic to make standardizing rule options logic a net reduction of complexity
+- This explicitly moves away from the JSON Schema / ajv style description of putting `default` values in `meta.schema`
```

Comment:
A potential downside of this is user confusion when one attempts to fruitlessly use the `default` schema property. I do think we should generally continue to accept any valid JSONSchema and allow users to annotate their schemas with `default` as desired, but this could become confusing if defaults end up out-of-sync between the schema and `defaultOptions`.

One potential mitigation is a new eslint-plugin-eslint-plugin rule `require-meta-default-options` which ensures that any `default` in the schema is represented in `defaultOptions` too. With an autofixer, this could also help with migration to the new property.

Reply:
I agree that any valid JSON Schema should be accepted. I'm less worried about developer confusion, though, as I think between documentation and tests, they will be able to figure out how to properly get defaults into their options.

Reply:
On the typescript-eslint side we already provide bespoke types for `meta.schema`, including [`default?`](https://github.com/typescript-eslint/typescript-eslint/blob/60df0bbafb6247341784dfbdf932585eac88b637/packages/utils/src/json-schema.ts#L152). We can always mark it `@deprecated` and then eventually remove it from the types altogether.

Reply:
I suppose you could do that, but that kind of goes against what I had suggested earlier which is that any JSONSchema should continue to be accepted. I assumed that ESLint/TSESlint wouldn't want to get into the business of disallowing some valid JSONSchemas.

Reply:
> I assumed that ESLint/TSESlint wouldn't want to get into the business of disallowing some valid JSONSchemas.

Quick note that the problem with "any valid JSON schema" is that by design it includes both well-formed and non-sensical JSON schemas.

For example one could write `{ type: 'string', additonalProperties: true }`. This is technically a valid schema but is clearly a programmer error.
A better example of this is that we ourselves had cases like this in our codebase: `{ type: 'object', properties: { ... }, additionalItems: false }`. Again this is technically a valid schema, but it is clearly a buggy schema - the author intended to user `additionalProperties`, not `additionalItems`.

In v6 (latest major) of our tooling we introduced the bespoke types that Josh linked above. These types are defined to allow "all well-formed, valid json schemas" - i.e. we now accept a subset of "valid json schemas".

Deprecating and banning `default` is something we can and would do, especially when this is implemented. We already recommend people don't use this - so pushing them away from it properly to the better, standardised `meta.defaultOptions` seems like a no-brainer!

Reply:
Fair enough. That reminds me that I myself had proposed that ESLint disallow no-op schemas in this RFC: https://github.com/eslint/rfcs/blob/main/designs/2021-schema-object-rules/README.md#no-op-schemas